Sources/Optimizer/stochastic_reconfiguration.hpp

```diff
@@ -19,143 +19,255 @@
+              bool use_cholesky = true, bool is_holomorphic = true)
+      __attribute__((deprecated))
+      : SR(use_cholesky ? LLT : ColPivHouseholder, diagshift, use_iterative,
+           is_holomorphic) {}
+
+  /**
+   * Solves the SR flow equation for the parameter update ẋ.
+   *
+   * The SR update is computed by solving the linear euqation
+   *    Sẋ = f
+   * where S is the covariance matrix of the partial derivatives
+   * O_i(v_j) = ∂/∂x_i log Ψ(v_j) and f is a generalized force (the loss
+   * gradient).
+   *
+   * @param Oks The matrix 𝕆 of centered log-derivatives,
+   *    𝕆_ij = O_i(v_j) - ⟨O_i⟩.
+   * @param grad The loss gradient f.
+   * @param deltaP Output parameter for the update ẋ.
+   */
+  void ComputeUpdate(OkRef Oks, GradRef grad, OutputRef deltaP);
+
+  void SetParameters(LSQSolver solver, double diagshift = 0.01,
+                     bool use_iterative = false, bool is_holomorphic = true);
+  void SetParameters(double diagshift = 0.01, bool use_iterative = false,
+                     bool use_cholesky = true, bool is_holomorphic = true)
+      __attribute__((deprecated));
+
+  /**
+   * Returns a string describing the current parameters of the SR class.
+   */
+  std::string GetInfoString();
+
+  /**
+   * Becca and Sorella (2017), pp. 143-144.
+   */
+  void SetScaleInvariantRegularization(bool enabled) {
+    if (use_iterative_ && enabled) {
+      // TODO: implement
+      throw std::runtime_error{
+          "Scale-invariant regularization is not implemented "
+          "for iterative solvers at the moment."};
+    }
+    if (enabled) {
+      InfoMessage() << "Using scale-invariant preconditioning." << std::endl;
+    }
+    scale_invariant_pc_ = enabled;
+  }
+  bool ScaleInvariantRegularizationEnabled() const {
+    return scale_invariant_pc_;
+  }
+
+  /**
+   * Returns the rank of the S matrix computed during the last call to
+   * `ComputeUpdate` or `nullopt`, in case storing the rank is not enabled
+   * and before the first call to `ComputeUpdate`.
+   *
+   * Storing the rank is enabled and disabled by `SetStoreRank` below.
+   */
+  nonstd::optional<Index> LastRank() { return last_rank_; }
+  bool StoreRankEnabled() const { return store_rank_; }
+  void SetStoreRank(bool enabled);
+
+  /**
+   * Returns a pointer to the full S matrix computed during the last call to
+   * `ComputeUpdate` or `nullptr`, in case storing the S matrix is not enabled
+   * and before the first call to `ComputeUpdate`.
+   *
+   * Storing the S matrix is enabled and disabled by `SetStoreFullSMatrix`
+   * below.
+   */
+  const MatrixXcd* LastSMatrix() const {
+    return last_S_.has_value() ? &last_S_.value() : nullptr;
+  }
+  bool StoreFullSMatrixEnabled() const { return store_full_S_matrix_; }
+  void SetStoreFullSMatrix(bool enabled);
+
+ private:
+  LSQSolver solver_;
   double sr_diag_shift_;
   bool use_iterative_;
-  bool use_cholesky_;
   bool is_holomorphic_;
 
+  bool scale_invariant_pc_ = false;
```

Comment:
Can we make this active by default and deactivate it automatically (for the moment) if the sparse solver is used?

Reply:
Sure.